### PR TITLE
[hue] Fix reconnection, parallel commands, trigger channels, and light level formula

### PR DESCRIPTION
**Pull request by andrewfg**
This PR fixes a number of issues raised since the initial release of API v2 support as follows:

1. Shortly before merging #13570 we discussed in this [thread](https://github.com/openhab/openhab-addons/pull/13570#discussion_r1241265369) some potential open issues concerning the handling of GO_AWAY events sent by the Hue bridge server. This PR resolves those issues.

2. When multiple HTTP PUT commands are sent over parallel HTTP2 streams, the Hue Bridge server will sometimes get confused, and return an HTML error. This PR resolves that issue by serializing the PUT commands.

3. A user reported a bug in the start up behaviour of  button trigger channels [here](https://community.openhab.org/t/philips-hue-clip-2-api-v2-discussion-thread/142111/166). When the binding is started up, it would wrongly activate the OH trigger channels for buttons with the last button event (button up). There was also a similar issue for rotary dials. This PR resolves those issues. Trigger channels are no longer activated on start up. However they are activated as before when actual new SSE events are received.

4. A user reported a bug with the formula for converting raw light level to Lux [here](https://community.openhab.org/t/philips-hue-clip-2-api-v2-discussion-thread/142111/178). This PR resolves that issue.

5. The Hue Bridge can sometimes send button events with a null 'last_event' field (probably occurs when a battery powered button comes and goes from the Zigbee net). This could cause an NPE. This PR resolves that issue.

### CODE REVIEW

**Comment by fwolter** [comment]
Looks good! Raise your hand when it's ready to merge.

**Comment by andrewfg** [comment]
> Raise your hand when it's ready to merge.

Nope. Unfortunately these changes make it worse  than before: now the bridge seems to go permanently offline every 24 hours at a specific time (which I had not observed before). So for the time being, please do NOT merge this PR. I will try to find a fix before M5, but as I am on vacation, I might not make it..

**Comment by andrewfg** [comment]
> now the bridge seems to go permanently offline every 24 hours at a specific time (which I had not observed before

This is weird. Every day at the same time (23:30), my bridge starts responding to commands from OH with an html error message rather than xml, which causes an “Unexpected Content-Type: text/html’” crash. It seems to be a bug in the bridge rather than the binding. Current work around is to disable/enable the binding, which works but I don’t know why. I need to dig further.

**Comment by andrewfg** [comment]
@fwolter / @lolodomo unfortunately I did not have time after returning from my vacation to finish this PR before v4.0 was released, however I think it is now ready to go into M1.

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java
Comment (jlaur) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:493]: Minor: This is exception from the constructor, not `initializeAssets()`. Also, strictly speaking, this is not a communication error, but failure to start a component (HttpClient).

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Button.java
Comment (jlaur) [inline bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/dto/clip2/Button.java:34]: Minor: There is no need to declare `IllegalArgumentException`. This is a runtime exception designed to fail fast, and should not normally be handled (but rather avoided).

**Review by jlaur — APPROVED** [review]
Thanks for the fixes! Just to be sure, how well tested are these fixes - are you running this version in your production system?

@fwolter - would it be possible for you to have a look also?

@openhab/add-ons-maintainers - I'm considering this for 4.0.1 today (hence the test question), would you agree?

**Comment by Novanic** [comment]
Hello,

I tried to migrate my Hue things to API v2, but noticed a few problems. Will this also get fixed with this issue?
- The dim switches are very slow. With API v1 my rules are reacting to the "dimmer_switch" (Number) channel. The rule switches my Hue light in <= 2 seconds. With the new API v2 things there is only a trigger channel. This switches my Hue light in between 2 and 6 seconds. That is annoying because it is far to slow.
- The API v2 migration has a bug. It migrates items from "dimmer_switch" (Number) channels to trigger channels. Trigger channels can not have linked items.
- sendCommand(1) isn't working anymore on API v2 "brightness" items. It doesn't switch the light on. That was working with API v1 things. sendCommand(10) and higher does still work.

I will stay on API v1 things for the moment.

Thank you and best regards

**Comment by jlaur** [comment]
> * sendCommand(1) isn't working anymore on API v2 "brightness" items. It doesn't switch the light on. That was working with API v1 things. sendCommand(10) and higher does still work.

Which channel is your "brightness" item linked to?

**Comment by Novanic** [comment]
> > * sendCommand(1) isn't working anymore on API v2 "brightness" items. It doesn't switch the light on. That was working with API v1 things. sendCommand(10) and higher does still work.
> 
> Which channel is your "brightness" item linked to?

The channel name/id is "brightness" (Dimmer).

**Comment by jlaur** [comment]
> @openhab/add-ons-maintainers - I'm considering this for 4.0.1 today (hence the test question), would you agree?

I'm away for the next two hours, so if you agree and the hotfix build is started today, please merge and cherry-pick this. Thanks!

**Review by fwolter — APPROVED** [review]
LGTM

**Comment by jlaur** [comment]
> > > * sendCommand(1) isn't working anymore on API v2 "brightness" items. It doesn't switch the light on. That was working with API v1 things. sendCommand(10) and higher does still work.
> > 
> > 
> > Which channel is your "brightness" item linked to?
> 
> The channel name/id is "brightness" (Dimmer).

I found in the code that there's a minimum dimming level you need to surpass to have it turned on:
https://github.com/openhab/openhab-addons/blob/f6e750a5ff1ad60ac83563db0d25aa5d18f02e9c/bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2ThingHandler.java#L393-L398

I don't know if this is exposed as a property, but it probably should be, so there is a way to check that. Please open a separate issue for tracking this.

**Comment by jlaur** [comment]
> * The dim switches are very slow. With API v1 my rules are reacting to the "dimmer_switch" (Number) channel. The rule switches my Hue light in <= 2 seconds. With the new API v2 things there is only a trigger channel. This switches my Hue light in between 2 and 6 seconds. That is annoying because it is far to slow.

This could be the same (or similar) issue as in #15164. You should be able to find information here to further track this problem which might not be caused by the Hue binding at all. You can also try to increase the number of threads in the thread pool. If you can provide information pointing to a problem in the Hue binding, please create a separate issue for this.

**Comment by jlaur** [comment]
> * The API v2 migration has a bug. It migrates items from "dimmer_switch" (Number) channels to trigger channels. Trigger channels can not have linked items.

Please create a separate issue for this.

File: bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java
Comment (andrewfg) [reply bundles/org.openhab.binding.hue/src/main/java/org/openhab/binding/hue/internal/handler/Clip2BridgeHandler.java:493]: > how well tested are these fixes

All the changes are backported to my v3.4.x operative system and therefore live tested.

**Comment by andrewfg** [comment]
> sendCommand(1) isn't working anymore on API v2 "brightness" items
> don't know if this is exposed as a property, but it probably should be

SendCommand works just fine. The lamp thing exposes a property `dimmingRange` and sendCommand() will only turn on the light if the commanded value is greater than or equal to the minimum `dimmingRange` for that particular lamp. Note that different lamp models have different minimum `dimmingRange` values, but typical minimum values can be 2% or more.

> The dim switches are very slow.

The bridge sends SSE event messages immediately. I you are seeing a delay between events and rules being triggered, then please submit trace logs showing your test cases.

As @jlaur said, this could be the same (or similar) issue as in https://github.com/openhab/openhab-addons/issues/15164.

Furthermore in my testing I have observed that if you simultaneously connect an API v1 Bridge Thing and an API v2 Bridge Thing to the same bridge, it might superficially appear as if both things are working. But in reality the bridge is struggling to handle both sets of API call in parallel and the bridge performance goes down the drain. Also there is no thread synchronization across mixed API v1 and API v2 Bridge things. Therefore my strong advice is to completely disable all API v1 things when you are testing API v2 functionality.

> It migrates items from "dimmer_switch" (Number) channels to trigger channels

This is intentional, and as specifically requested by OH maintainers. Initially I had coded the button channels (and rotary wheel channels) as state channels of type Number. However @kaikreuzer quite rightly pointed out that those channels do not have a persistent state, and quite rightly asked them to be exposed as trigger channels. Which I did. 

PS furthermore, reverting to a number state channel type would conflict with the conclusion already reached [here](https://community.openhab.org/t/philips-hue-clip-2-api-v2-discussion-thread/142111/170?u=andrewfg) and resolved in this PR.

> Trigger channels can not have linked items.

@Novanic your statement is wrong: of course you can link an item to a trigger channel; you just need to remember that trigger channel values are expressed as String rather then Number.

image

**Pull request merged**
